modules/build/src/main/scala/scala/build/Build.scala

```diff
@@ -149,117 +150,113 @@ object Build {
     val sharedOptions = crossSources.sharedOptions(options)
     val crossOptions  = sharedOptions.crossOptions
 
+    def doPostProcess(build: Build, inputs: Inputs, scope: Scope): Unit = build match {
+      case build: Build.Successful =>
+        postProcess(
+          build.generatedSources,
+          inputs.generatedSrcRoot(scope),
+          build.output,
+          logger,
+          inputs.workspace,
+          updateSemanticDbs = true,
+          scalaVersion = build.project.scalaCompiler.scalaVersion
+        ).left.foreach(_.foreach(logger.message(_)))
+      case _ =>
+    }
+
     def doBuild(
-      overrideOptions: BuildOptions,
-      scope: Scope
-    ): Either[BuildException, Build] = either {
+      overrideOptions: BuildOptions
+    ): Either[BuildException, (Build, Build)] = either {
 
       val inputs0 = updateInputs(
         inputs,
         overrideOptions.orElse(options) // update hash in inputs with options coming from the CLI or cross-building, not from the sources
       )
 
-      val baseOptions = overrideOptions.orElse(sharedOptions)
+      def doBuildScope(options: BuildOptions, scope: Scope): Either[BuildException, Build] =
+        either {
+          val crossSources0 = crossSources.withVirtualDir(inputs0, scope, options)
+          val scopedSources = value(crossSources0.scopedSources(options))
+          val sources       = scopedSources.sources(scope, options)
+
+          val generatedSources =
+            sources.generateSources(inputs0.generatedSrcRoot(scope))
+          val buildOptions = sources.buildOptions
+
+          val scopeParams =
+            if (scope == Scope.Main) Nil
+            else Seq(scope.name)
+
+          buildClient.setProjectParams(scopeParams ++ value(buildOptions.projectParams))
+
+          val res = build(
+            inputs0,
+            sources,
+            inputs0.generatedSrcRoot(scope),
+            generatedSources,
+            buildOptions,
+            scope,
+            logger,
+            buildClient,
+            bloopServer
+          )
 
-      val crossSources0 = crossSources.withVirtualDir(inputs0, scope, baseOptions)
+          value(res)
+        }
 
-      val scopedSources = value(crossSources0.scopedSources(baseOptions))
-      val sources       = scopedSources.sources(scope, baseOptions)
+      val mainOptions = overrideOptions.orElse(sharedOptions)
+      val mainBuild   = value(doBuildScope(mainOptions, Scope.Main))
 
-      val generatedSources = sources.generateSources(inputs0.generatedSrcRoot(scope))
-      val buildOptions     = sources.buildOptions
+      val testBuild = value {
+        mainBuild match {
+          case s: Build.Successful =>
+            val extraTestOptions = BuildOptions(
+              classPathOptions = ClassPathOptions(
+                extraClassPath = Seq(s.output)
+              )
+            )
+            val testOptions = mainOptions.orElse(extraTestOptions)
```

Comment:
I would have done it the other way around here
```suggestion
            val testOptions = extraTestOptions.orElse(mainOptions)
```
so that things in `extraTestOptions` have a higher priority (that would put `s.output` first in `extraClassPath`).